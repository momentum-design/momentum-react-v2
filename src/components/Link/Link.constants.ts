const CLASS_PREFIX = 'md-link';

const DEFAULTS = {
  DISABLED: false,
  INVERTED: false,
};

const STYLE = {
  text: `${CLASS_PREFIX}-text`,
  wrapper: `${CLASS_PREFIX}-wrapper`,
  container: `${CLASS_PREFIX}-container`,
  icon: `${CLASS_PREFIX}-icon`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
