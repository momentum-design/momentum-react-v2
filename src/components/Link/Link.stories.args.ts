import {
  commonAriaLink,
  commonHTMLAttributes,
  commonStyles,
} from '../../storybook/helper.stories.argtypes';

import { Link_CONSTANTS as CONSTANTS } from './';

export default {
  ...commonStyles,
  ...commonHTMLAttributes,
  ...commonAriaLink,
  children: {
    description: 'Provides the child nodes for this element.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  disabled: {
    description: 'Whether to render the `<Link />` as disabled.',
    options: [true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.DISABLED,
      },
    },
  },
  inverted: {
    description: 'Whether to render the `<Link />` with inverted theme colors',
    options: [true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.INVERTED,
      },
    },
  },
  hasExternalLinkIcon: {
    description: 'whether this component has icon `name = pop-out` in the right',
    options: [true, false],
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: false,
      },
    },
  },
  externalLinkIconProps: {
    description: `IconProps (scale and autoscale only) for the 'pop-out' icon which is shown on the right side of the link.`,
    control: { type: 'object' },
    table: {
      type: {
        summary: 'object',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  tooltipContent: {
    description:
      'Used for tooltip content that gets triggered when the link is hovered; e.g Opens a new tab',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  tooltipType: {
    description: `Determines, whether the tooltip is the description or the label of the trigger component, or none`,
    options: ['none', 'label', 'description'],
    control: { type: 'select' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'description',
      },
    },
  },
  tabIndex: {
    description: 'tabIndex for `<a>` tag',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
};
