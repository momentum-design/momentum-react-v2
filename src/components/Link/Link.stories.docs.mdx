The `<Link />` component.

## React Aria Link Support

This element utilizes `AriaLinkprops` for the `useLink()`
[React Aria](https://react-spectrum.adobe.com/react-aria/useLink.html) method. This allows for the
usage of `LinkProps`、`LinkDOMProps`、`AriaLabelingProps`.

`AriaLinkProps` allows for the usage of common #aria attributes, such as
`aria-label`、`aria-details` 、`aria-labelledby`、`aria-describedby`.
