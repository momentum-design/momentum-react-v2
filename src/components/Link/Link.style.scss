.md-link-wrapper {
  border: var(--md-globals-border-clear);
  color: var(--mds-color-theme-text-accent-normal);
  cursor: pointer;
  outline: none !important;
  padding: 0;
  text-decoration: underline;
  transition: background-color 0.2s, color 0.2s, border-color 0.2s;
  display: inline-block;

  // have based a:focus style on momentum-ui.css.
  // need to overwrite focus style and src
  &:focus,
  &.focus {
    color: var(--mds-color-theme-text-accent-normal);
  }

  &:hover,
  &.hover {
    color: var(--mds-color-theme-text-accent-hover);
  }

  &:active,
  &.active {
    color: var(--mds-color-theme-text-accent-active);
  }

  &[data-inverted='true'] {
    color: var(--mds-color-theme-inverted-text-accent-normal);

    &:focus,
    &.focus {
      color: var(--mds-color-theme-inverted-text-accent-normal);
    }

    &:hover,
    &.hover {
      color: var(--mds-color-theme-inverted-text-accent-hover);
    }

    &:active,
    &.active {
      color: var(--mds-color-theme-inverted-text-accent-hover);
    }

    &[data-disabled='true'],
    &.disable {
      color: var(--mds-color-theme-text-primary-disabled);
      cursor: auto;
    }
  }

  &[data-disabled='true'],
  &.disable {
    color: var(--mds-color-theme-text-primary-disabled);
    cursor: auto;
  }
}

.md-link-container {
  display: flex;
  align-items: center;

  .md-link-icon {
    margin-left: 0.3125rem;
  }
}
