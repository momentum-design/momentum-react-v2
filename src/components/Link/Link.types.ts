import { ReactNode } from 'react';
import { AriaLinkProps } from '@react-types/link';
import { Props as IconProps } from '../Icon/Icon.types';

export interface Props extends AriaLinkProps {
  /**
   * Custom class to be able to override the component's CSS
   */
  className?: string;
  /**
   * Child components of this component. This prop replaces primaryText.
   */
  children?: ReactNode;

  /**
   * Whether or not this component is disabled.
   */
  disabled?: boolean;

  /**
   * Whether or not this component has inverted theme colors.
   */
  inverted?: boolean;

  /**
   * title to use for this component.
   */
  title?: string;

  /**
   * whether or not this component has icon【name = pop-out】 in the right
   */
  isWithIcon?: boolean;

  /**
   * icon props
   * When use iconProps isWithIcon is necessary
   */
  iconProps?: Partial<IconProps>;
}
