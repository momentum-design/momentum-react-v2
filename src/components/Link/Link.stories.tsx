import { MultiTemplateWithPseudoStates, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import Link, { LinkProps } from './';
import argTypes from './Link.stories.args';
import Documentation from './Link.stories.docs.mdx';
import React from 'react';
import { omit } from 'lodash';

export default {
  title: 'Momentum UI/Link',
  component: Link,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
  args: {
    children: 'Example Text',
  },
};

const Example = Template<LinkProps>(Link).bind({});

Example.args = {
  href: 'https://www.webex.com/',
  target: '_blank',
};

Example.argTypes = { ...argTypes };

const LinkContainer = (props) => {
  if (props.inverted) {
    return (
      <div style={{ backgroundColor: 'var(--mds-color-theme-inverted-background-normal)' }}>
        <Link {...props} />
      </div>
    );
  } else {
    return <Link {...props} />;
  }
};

const States = MultiTemplateWithPseudoStates<LinkProps>(LinkContainer).bind({});

States.parameters = {
  variants: [
    { label: 'Default link', href: 'https://www.webex.com/' },
    { label: 'Inverted', inverted: true, href: 'https://www.webex.com/' },
  ],
};

States.argTypes = { ...argTypes };
States.argTypes = omit(States.argTypes, ['disabled', 'inverted']);

const LinkHasIcon = Template<LinkProps>(Link).bind({});

LinkHasIcon.args = {
  href: 'https://www.webex.com/',
  target: '_blank',
  tooltipContent: 'opens a new window',
  tooltipType: 'description',
};

LinkHasIcon.argTypes = { ...argTypes };

export { Example, States, LinkHasIcon };
