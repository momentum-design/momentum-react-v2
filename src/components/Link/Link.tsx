import React, { forwardRef, useRef, RefObject } from 'react';
import FocusRing from '../FocusRing';
import { useLink } from '@react-aria/link';
import { DEFAULTS, STYLE } from './Link.constants';
import { Props } from './Link.types';
import './Link.style.scss';
import classnames from 'classnames';
import Icon from '../Icon';
import Tooltip from '../Tooltip';

const Link = forwardRef((props: Props, providedRef: RefObject<HTMLAnchorElement>) => {
  const {
    className,
    title,
    hasIcon,
    iconProps,
    tooltipContent,
    tooltipType,
    disabled,
    inverted,
    style,
    ...otherProps
  } = props;
  const internalRef = useRef();
  const ref = providedRef || internalRef;

  const mutatedProps = {
    ...otherProps,
    title,
    isDisabled: disabled,
    isInverted: inverted,
  };

  // prevent change the window.opener.location to some phishing page
  const isExternalLinkl = props.target === '_blank';

  const { linkProps } = useLink({ ...mutatedProps, elementType: 'a' }, ref);

  const commonProps = {
    ...linkProps,
    style,
    ref: ref,
    title: title,
    rel: isExternalLinkl ? 'noopener noreferrer' : '',
    className: classnames(STYLE.wrapper, className),
    'data-disabled': disabled || DEFAULTS.DISABLED,
    'data-inverted': inverted || DEFAULTS.INVERTED,
  };

  return (
    <FocusRing disabled={disabled}>
      <>
        {tooltipContent && (
          <Tooltip
            type={tooltipType}
            placement="bottom"
            triggerComponent={
              <a {...commonProps}>
                <div className={STYLE.container}>
                  {props.children}
                  {hasIcon && (
                    <Icon className={STYLE.icon} scale={16} name="pop-out" {...iconProps} />
                  )}
                </div>
              </a>
            }
          >
            {tooltipContent}
          </Tooltip>
        )}
        {!tooltipContent && <a {...commonProps}>{props.children}</a>}
      </>
    </FocusRing>
  );
});

Link.displayName = 'Link';

export default Link;
