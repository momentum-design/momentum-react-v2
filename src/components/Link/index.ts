import { default as Link } from './Link';
import * as CONSTANTS from './Link.constants';
import { Props } from './Link.types';

export { CONSTANTS as Link_CONSTANTS };

export type LinkProps = Props;

export default Link;
