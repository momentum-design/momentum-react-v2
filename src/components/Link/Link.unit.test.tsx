import React from 'react';
import { mount } from 'enzyme';
import { DEFAULTS, STYLE } from './Link.constants';
import { IconNext, LinkNext, TooltipNext } from '@momentum-ui/react-collaboration';
import { mountAndWait, triggerPress } from '../../../test/utils';

describe('Link', () => {
  describe('snapshot', () => {
    let container;

    it('should match snapshot', () => {
      expect.assertions(1);

      container = mount(<LinkNext>LinkNext</LinkNext>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when disabled', () => {
      expect.assertions(1);

      const disabled = !DEFAULTS.DISABLED;

      container = mount(<LinkNext disabled={disabled}>LinkNext</LinkNext>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when inverted', () => {
      expect.assertions(1);

      const inverted = !DEFAULTS.INVERTED;

      container = mount(<LinkNext inverted={inverted}>Hyperlink</LinkNext>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with title', () => {
      expect.assertions(1);

      const title = 'Example Text';

      const container = mount(<LinkNext title={title} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its main class', () => {
      expect.assertions(1);

      const element = mount(<LinkNext />)
        .find(LinkNext)
        .getDOMNode();

      expect(element.classList.contains(STYLE.wrapper)).toBe(true);
    });

    it('should have custom class if provided', async () => {
      const testClass = 'testClass';

      const wrapper = await mount(<LinkNext className={testClass} />);

      const element = wrapper.find('a').getDOMNode();

      expect(element.classList.contains(STYLE.wrapper)).toBe(true);
      expect(element.classList.contains(testClass)).toBe(true);
    });

    it('should pass disabled prop', async () => {
      expect.assertions(1);

      const disabled = !DEFAULTS.DISABLED;
      const wrapper = await mountAndWait(<LinkNext disabled={disabled} />);

      const element = wrapper.find('a').getDOMNode();

      expect(element.getAttribute('data-disabled')).toBe(`${disabled}`);
    });

    it('should pass inverted prop', async () => {
      expect.assertions(1);

      const inverted = !DEFAULTS.INVERTED;
      const wrapper = await mountAndWait(<LinkNext inverted={inverted} />);

      const element = wrapper.find('a').getDOMNode();

      expect(element.getAttribute('data-inverted')).toBe(`${inverted}`);
    });

    it('should pass child prop', async () => {
      expect.assertions(1);

      const child = '1';

      const wrapper = await mountAndWait(<LinkNext>{child}</LinkNext>);

      const element = wrapper.find('a').getDOMNode();

      expect(element.innerHTML).toBe(child);
    });

    it('should have provided title when title is provided', async () => {
      expect.assertions(1);

      const title = 'Example Text';

      const wrapper = await mountAndWait(<LinkNext title={title} />);

      const element = wrapper.find('a').getDOMNode();

      expect(element.getAttribute('title')).toBe(title);
    });

    it('should have Tooltip when tooltipContent is provided', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(<LinkNext tooltipContent={tooltipContent} />);

      const element = wrapper.find(TooltipNext);

      expect(element.props()).toEqual({
        type: undefined,
        placement: 'bottom',
        triggerComponent: expect.any(Object),
        children: tooltipContent,
      });
    });

    it('should have Tooltip when tooltipContent and tooltipType is provided', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(
        <LinkNext tooltipContent={tooltipContent} tooltipType={'description'} />
      );

      const element = wrapper.find(TooltipNext);

      expect(element.props()).toEqual({
        type: 'description',
        placement: 'bottom',
        triggerComponent: expect.any(Object),
        children: tooltipContent,
      });
    });

    it('should have icon when hasExternalLinkIcon is provided', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(
        <LinkNext tooltipContent={tooltipContent} hasExternalLinkIcon={true} />
      );

      const element = wrapper.find(IconNext);

      expect(element.props()).toEqual({
        className: STYLE.icon,
        scale: 16,
        name: 'pop-out',
      });
    });

    it('should have icon when target is _blank', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(
        <LinkNext tooltipContent={tooltipContent} target="_blank" />
      );

      const element = wrapper.find(IconNext);

      expect(element.props()).toEqual({
        className: STYLE.icon,
        scale: 16,
        name: 'pop-out',
      });
    });

    it('should not have icon when target is _blank and hasExternalLinkIcon is false', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(
        <LinkNext tooltipContent={tooltipContent} target="_blank" hasExternalLinkIcon={false} />
      );

      expect(wrapper.find(IconNext)).toHaveLength(0);
    });

    it('should have icon props when externalLinkIconProps is provided', async () => {
      expect.assertions(1);

      const tooltipContent = 'open a new window';

      const wrapper = await mountAndWait(
        <LinkNext
          tooltipContent={tooltipContent}
          hasExternalLinkIcon={true}
          externalLinkIconProps={{ autoScale: true }}
        />
      );

      const element = wrapper.find(IconNext);

      expect(element.props()).toEqual({
        className: STYLE.icon,
        scale: 16,
        name: 'pop-out',
        autoScale: true,
      });
    });

    it('should have rel when target is provided', async () => {
      expect.assertions(1);

      const wrapper = await mountAndWait(<LinkNext target={'_blank'} />);

      const element = wrapper.find('a');

      expect(element.props()).toEqual({
        rel: 'noopener noreferrer',
        children: undefined,
        className: 'md-link-wrapper',
        'data-disabled': false,
        'data-inverted': false,
        target: '_blank',
        title: undefined,
      });
    });

    it('should have tabIndex when tabIndex is provided', async () => {
      expect.assertions(1);

      const wrapper = await mountAndWait(<LinkNext tabIndex={0}/>);

      const element = wrapper.find('a');

      expect(element.props()).toEqual({
        children: undefined,
        className: 'md-link-wrapper',
        'data-disabled': false,
        'data-inverted': false,
        tabIndex: 0,
        style: undefined,
        onClick: undefined,
        title: undefined,
        rel: '',
      });
    });
  });

  describe('actions', () => {
    it('should handle mouse press events', () => {
      expect.assertions(2);
      const mockCallback = jest.fn();
      const component = mount(<LinkNext onPress={mockCallback} />).find(LinkNext);
      component.props().onPress({
        type: 'press',
        pointerType: 'mouse',
        altKey: false,
        shiftKey: false,
        ctrlKey: false,
        metaKey: false,
        target: component.getDOMNode(),
        continuePropagation: expect.any(Function),
      });

      expect(mockCallback).toBeCalledTimes(1);

      triggerPress(component);

      expect(mockCallback).toBeCalledTimes(2);
    });
  });
});
