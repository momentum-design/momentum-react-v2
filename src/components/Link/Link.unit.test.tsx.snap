// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link snapshot should match snapshot 1`] = `
<Link>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-link-wrapper"
        data-disabled={false}
        data-inverted={false}
        elementType="a"
        onPress={[MockFunction]}
      >
        LinkNext
      </a>
    </FocusRing>
  </FocusRing>
</Link>
`;

exports[`Link snapshot should match snapshot when disabled 1`] = `
<Link
  disabled={true}
>
  <FocusRing
    disabled={true}
  >
    <FocusRing
      disabled={true}
      focusClass="md-focus-ring-wrapper-disabled"
      focusRingClass="md-focus-ring-wrapper-disabled"
    >
      <a
        className="md-link-wrapper"
        data-disabled={true}
        data-inverted={false}
        elementType="a"
        isDisabled={true}
        onPress={[MockFunction]}
      >
        LinkNext
      </a>
    </FocusRing>
  </FocusRing>
</Link>
`;

exports[`Link snapshot should match snapshot when inverted 1`] = `
<Link
  inverted={true}
>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-link-wrapper"
        data-disabled={false}
        data-inverted={true}
        elementType="a"
        inverted={true}
        isInverted={true}
        onPress={[MockFunction]}
      >
        Hyperlink
      </a>
    </FocusRing>
  </FocusRing>
</Link>
`;

exports[`Link snapshot should match snapshot with title 1`] = `
<Link
  title="Example Text"
>
  <FocusRing>
    <FocusRing
      focusClass="md-focus-ring-wrapper children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <a
        className="md-link-wrapper"
        data-disabled={false}
        data-inverted={false}
        elementType="a"
        onPress={[MockFunction]}
        title="Example Text"
      />
    </FocusRing>
  </FocusRing>
</Link>
`;
